src/Type/UnionType.php

```diff
@@ -102,22 +123,54 @@ private function getSortedTypes(): array
 	 */
 	public function getReferencedClasses(): array
 	{
-		return UnionTypeHelper::getReferencedClasses($this->getTypes());
+		$classes = [];
+		foreach ($this->types as $type) {
+			$classes[] = $type->getReferencedClasses();
+		}
+
+		return array_merge(...$classes);
 	}
 
 	public function getArrays(): array
 	{
-		return UnionTypeHelper::getArrays($this->getTypes());
+		$arrays = [];
+		foreach ($this->types as $type) {
+			$arrays[] = $type->getArrays();
+		}
+
+		if ($arrays === []) {
+			return [];
+		}
+
+		return array_merge(...$arrays);
 	}
 
 	public function getConstantArrays(): array
 	{
-		return UnionTypeHelper::getConstantArrays($this->getTypes());
+		$constantArrays = [];
+		foreach ($this->getTypesOfClass(ConstantArrayType::class) as $type) {
+			$constantArrays[] = $type->getConstantArrays();
+		}
+
+		if ($constantArrays === []) {
+			return [];
+		}
+
+		return array_merge(...$constantArrays);
 	}
 
 	public function getConstantStrings(): array
 	{
-		return UnionTypeHelper::getConstantStrings($this->getTypes());
+		$strings = [];
+		foreach ($this->getTypesOfClass(ConstantStringType::class) as $type) {
```

Comment:
Same logic as with getConstantArrays.

You could test this with:

* `''|numeric-string` - should return empty array
* Create a fake denormalized type and test that `getConstantStrings` returns an array of 'foo' and 'bar':

```php
new UnionType([
    new IntersectionType([new ConstantStringType('foo'), new AccessoryLiteralStringType()]),
    new IntersectionType([new ConstantStringType('bar'), new AccessoryLiteralStringType()]),
])
```

(In the future there could be a valid intersection like this - for example if we add nominal types and people are able to define new types like `email-string`, we could have `'ondrej@mirtes.cz'&email-string`.
```

Reply:
thx for the detailled feedback. I couldn't have come up with such tests on my own.